src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -0,0 +1,245 @@
+        private const string EncodingNameGzip = "GZIP";
+        private const string HeaderNameContentEncoding = "Content-Encoding";
+        private const string HeaderNameContentLength = "Content-Length";
+        private const string HeaderNameSetCookie = "Set-Cookie";
+        private static readonly string[] s_HttpHeadersSeparator = { "\r\n" };
+
+        public static HttpResponseMessage CreateResponseMessage(
+            WinHttpRequestState state,
+            bool doManualDecompressionCheck)
+        {
+            HttpRequestMessage request = state.RequestMessage;
+            SafeWinHttpHandle requestHandle = state.RequestHandle;
+            CookieUsePolicy cookieUsePolicy = state.Handler.CookieUsePolicy;
+            CookieContainer cookieContainer = state.Handler.CookieContainer;
+            var response = new HttpResponseMessage();
+            bool useDeflateDecompression = false;
+            bool useGzipDecompression = false;
+
+            // Get HTTP version, status code, reason phrase from the response headers.
+            string version = GetResponseHeaderStringInfo(requestHandle, Interop.WinHttp.WINHTTP_QUERY_VERSION);
+            if (string.Compare("HTTP/1.1", version, StringComparison.OrdinalIgnoreCase) == 0)
+            {
+                response.Version = new Version(1, 1);
+            }
+            else if (string.Compare("HTTP/1.0", version, StringComparison.OrdinalIgnoreCase) == 0)
+            {
+                response.Version = new Version(1, 0);
+            }
+            else
+            {
+                response.Version = null;
+            }
+
+            response.StatusCode = (HttpStatusCode)GetResponseHeaderNumberInfo(
+                requestHandle,
+                Interop.WinHttp.WINHTTP_QUERY_STATUS_CODE);
+            response.ReasonPhrase = GetResponseHeaderStringInfo(
+                requestHandle,
+                Interop.WinHttp.WINHTTP_QUERY_STATUS_TEXT);
+
+            if (doManualDecompressionCheck)
+            {
+                string contentEncoding = GetResponseHeaderStringInfo(
+                    requestHandle,
+                    Interop.WinHttp.WINHTTP_QUERY_CONTENT_ENCODING);
+                if (!string.IsNullOrEmpty(contentEncoding))
+                {
+                    if (contentEncoding.IndexOf(EncodingNameDeflate, StringComparison.OrdinalIgnoreCase) > -1)
+                    {
+                        useDeflateDecompression = true;
+                    }
+                    else if (contentEncoding.IndexOf(EncodingNameGzip, StringComparison.OrdinalIgnoreCase) > -1)
+                    {
+                        useGzipDecompression = true;
+                    }
+                }
+            }
+
+            // Create response stream and wrap it in a StreamContent object.
+            var responseStream = new WinHttpResponseStream(state);
+            Stream decompressedStream = responseStream;
```

Comment:
Seems like if you were to move these two lines to above the previous if block, then you wouldn't need useDeflateDecompression/useGzipDecompression nor the subsequent if block and could just construct the appropriate decompression stream in the above if block.